### PR TITLE
Change operations a bit + add TransposedConv2D layer

### LINKED ISSUE #53: Implement Convolution Transpose Layer
While a Convolutional layer already is implemented in the existing layers, if we had a transposed version we could start to build up interesting architectures such as GANs and VAEs.

### PR DESCRIPTION
**Pull request by tanmayb123**
Closes #53

### CODE REVIEW

**Comment by tanmayb123** [comment]
I've tested this out locally, and everything seems to be working. @rxwei - do you have some feedback on how this implementation could be made better?

File: Sources/DeepLearning/Operators.swift [hunk 1]
```diff
@@ -120,91 +120,78 @@ public extension Padding {
...
+                self.conv2DBackpropInput(shape: filterSizes, filter: v, strides: strides,
+                                              padding: padding),
                 input.convolved2D(withFilter: v, strides: strides, padding: padding)
             )
         })
     }
-
+    
     @inlinable
     func _vjpConvolved2D(
```
Comment (rxwei) [inline Sources/DeepLearning/Operators.swift]: The extension is public, but all VJP functions should be explicitly `internal`.

Comment (tanmayb123) [reply Sources/DeepLearning/Operators.swift]: Done.

File: Sources/DeepLearning/Operators.swift [hunk 2]
```diff
@@ -269,9 +256,11 @@ extension Tensor where Scalar: TensorFlowFloatingPoint {
             )
         })
     }
+
```
Comment (rxwei) [inline Sources/DeepLearning/Operators.swift]: Remove unnecessary empty line.

File: Sources/DeepLearning/Operators.swift [hunk 3]
```diff
@@ -269,9 +256,11 @@ extension Tensor where Scalar: TensorFlowFloatingPoint {
             )
         })
     }
+    
 }
 
 public extension Tensor where Scalar: FloatingPoint {
+
```
Comment (rxwei) [inline Sources/DeepLearning/Operators.swift]: Remove unnecessary empty line. The Google Swift Style Guide neither encourages or discourages the inclusion of an empty line in the beginning and the end of a type declaration, but the choice in our code bases is to not to add them. Same for other places.

File: Sources/DeepLearning/Operators.swift [hunk 4]
```diff
@@ -120,93 +120,80 @@ public extension Padding {
...
         _ filter: Tensor,
-        _ backpropOutput: Tensor,
         _ strides: (Int32, Int32, Int32, Int32),
         _ padding: Padding
     ) -> (Tensor, (Tensor) -> (Tensor, Tensor)) {
-        let value = _TFConv2DBackpropInput(shape: shape, filter: filter,
-                                           backpropOutput: backpropOutput,
-                                           strides: strides, padding: padding)
+        let value = conv2DBackpropInput(shape: shape, filter: filter, strides: strides,
+                                             padding: padding)
```
Comment (rxwei) [inline Sources/DeepLearning/Operators.swift]: Align argument.

File: Sources/DeepLearning/Operators.swift [hunk 5]
```diff
@@ -120,93 +120,80 @@ public extension Padding {
...
         return (value, { v in
             return (
-                self._TFConv2DBackpropFilter(input: v, filterSizes: shape,
-                                             backpropOutput: backpropOutput,
-                                             strides: strides, padding: padding),
+                self.conv2DBackpropFilter(input: v, filterSizes: shape, strides: strides,
+                                               padding: padding),
```
Comment (rxwei) [inline Sources/DeepLearning/Operators.swift]: Align argument.

File: Sources/DeepLearning/Operators.swift [hunk 6]
```diff
@@ -120,93 +120,80 @@ public extension Padding {
...
         _ filterSizes: Tensor<Int32>,
-        _ backpropOutput: Tensor,
         _ strides: (Int32, Int32, Int32, Int32),
         _ padding: Padding
     ) -> (Tensor, (Tensor) -> (Tensor, Tensor)) {
-        let value = _TFConv2DBackpropFilter(input: input, filterSizes: filterSizes,
-                                            backpropOutput: backpropOutput,
-                                            strides: strides, padding: padding)
+        let value = conv2DBackpropFilter(input: input, filterSizes: filterSizes,
+                                              strides: strides, padding: padding)
```
Comment (rxwei) [inline Sources/DeepLearning/Operators.swift]: Align argument.

File: Sources/DeepLearning/Operators.swift [hunk 7]
```diff
@@ -120,93 +120,80 @@ public extension Padding {
...
         return (value, { v in
             return (
-                self._TFConv2DBackpropInput(shape: filterSizes, filter: v,
-                                            backpropOutput: backpropOutput,
-                                            strides: strides, padding: padding),
+                self.conv2DBackpropInput(shape: filterSizes, filter: v, strides: strides,
+                                              padding: padding),
```
Comment (rxwei) [inline Sources/DeepLearning/Operators.swift]: Align argument.

File: Sources/DeepLearning/Operators.swift [hunk 8]
```diff
@@ -120,93 +120,80 @@ public extension Padding {
...
         return Raw.conv2DBackpropFilter(
             input,
             filterSizes: filterSizes,
-            outBackprop: backpropOutput,
+            outBackprop: self,
             strides: [strides.0, strides.1, strides.2, strides.3],
             padding: padding.raw)
     }
-
+
```
Comment (rxwei) [inline Sources/DeepLearning/Operators.swift]: Remove trailing whitespace.

File: Sources/DeepLearning/Operators.swift [hunk 9]
```diff
@@ -120,93 +120,80 @@ public extension Padding {
...
+    
     /// TensorFlow builtin conv2d gradient helper for the input.
     @inlinable
-    @differentiable(
-        wrt: (filter, backpropOutput),
-        vjp: _vjpTFConv2DBackpropInput(_:_:_:_:_:)
-    )
-    func _TFConv2DBackpropInput(
+    @differentiable(wrt: (self, filter), vjp: _vjpConv2DBackpropInput)
+    func conv2DBackpropInput(
```
[the remark this answers is not in the document]
Comment (tanmayb123) [reply Sources/DeepLearning/Operators.swift]: `conv2DBackpropInput` is needed for `TransposedConv2D`, so I can't make it internal. I can make `conv2DBackpropFilter` internal though. What do you think?

[the remark this answers is not in the document]
Comment (tanmayb123) [reply Sources/DeepLearning/Operators.swift]: Unless you want me to make a further wrapper around this function called `transposeConvolved2D` which handles shape calculation + calling this function.

[the remark this answers is not in the document]
Comment (rxwei) [reply Sources/DeepLearning/Operators.swift]: Hmm, I don't understand. Why is it the case you can't make it `internal`?

[the remark this answers is not in the document]
Comment (tanmayb123) [reply Sources/DeepLearning/Operators.swift]: Ah sorry, I misunderstood in that case.

**Comment by tanmayb123** [comment]
I believe I've fixed all the above items from your review.

File: Sources/DeepLearning/Operators.swift [hunk 10]
```diff
@@ -298,7 +284,7 @@ public extension Tensor where Scalar: FloatingPoint {
             strides: [strides.0, strides.1, strides.2, strides.3],
             padding: padding.raw)
     }
-
+
```
Comment (rxwei) [inline Sources/DeepLearning/Operators.swift]: Remove trailing whitespace.

**Comment by rxwei** [comment]
Could you fix the test failure?

**Comment by tanmayb123** [comment]
It worked.

**Comment by rxwei** [comment]
Thanks!

**Pull request merged**